python/cudf/cudf/core/udf/groupby_utils.py

```diff
@@ -147,12 +149,18 @@ def jit_groupby_apply(offsets, grouped_values, function, *args):
     offsets = cp.asarray(offsets)
     ngroups = len(offsets) - 1
 
-    kernel, return_type = _get_groupby_apply_kernel(
-        grouped_values, function, args
-    )
-    return_type = numpy_support.as_dtype(return_type)
+    cache_key = _generate_cache_key(grouped_values, function)
+    if precompiled.get(cache_key) is not None:
+        kernel, return_type = precompiled[cache_key]
+    else:
+        kernel, return_type = _get_groupby_apply_kernel(
+            grouped_values, function, args
+        )
+        precompiled[cache_key] = (kernel, return_type)
```

Comment:
Fill the cache if it's empty, then read the cache. This reduces duplication.
```suggestion
    if cache_key not in precompiled:
        precompiled[cache_key] = _get_groupby_apply_kernel(
            grouped_values, function, args
        )
    kernel, return_type = precompiled[cache_key]
```